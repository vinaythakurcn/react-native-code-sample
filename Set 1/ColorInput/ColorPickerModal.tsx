import React, { useRef, useState } from 'react'
import { Modal, Dimensions, StyleSheet } from 'react-native'
import WheelColorPicker from 'react-native-wheel-color-picker'
import styled from 'styled-components/native'
import Colors from '../../../theme/colors'

import Button from '../../FormInput/Button'

export interface IColorPickerModalProps {
    isOpen: boolean
    value?: string
    onChange: (value: string) => void
    onClose: () => void
}

const { height } = Dimensions.get('window')

const ModalContainer = styled.SafeAreaView`
    height: ${height}px;
    background-color: ${Colors.white};
`

const ContentArea = styled.View`
    height: 100%;
    margin: 12px 18px;
`
const FooterActionArea = styled.View`
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    align-items: center;
    margin-bottom: 12px;
`
const PickerArea = styled.ScrollView`
    flex: 1;
`

const ColorPickerModal: React.FC<IColorPickerModalProps> = (props) => {
    const { isOpen, onClose, onChange, value } = props
    const [currentValue, setCurrentValue] = useState<string>(value || '')
    const wheelRef = useRef<any>()

    return (
        <Modal
            animationType='slide'
            transparent={false}
            visible={isOpen}
            onRequestClose={() => {
                onClose()
            }}
        >
            <ModalContainer>
                <ContentArea>
                    <PickerArea>
                        <WheelColorPicker
                            ref={(r) => {
                                wheelRef.current = r
                            }}
                            color={value || Colors.requestFailedColor}
                            onColorChangeComplete={(color) => setCurrentValue(color)}
                        />
                    </PickerArea>
                    <FooterActionArea>
                        <Button
                            variant='default'
                            onPress={() => onClose()}
                            title={'Cancel'}
                            style={{ width: '45%', marginRight: 8 }}
                            icon={'highlight_off'}
                            IconStyle={style.btnStyle}
                        />
                        <Button
                            variant='default'
                            onPress={() => onChange(currentValue)}
                            title={'Apply'}
                            style={{ backgroundColor: Colors.darkButtonHover, borderWidth: 0, width: '45%', marginLeft: 8 }}
                            textStyle={{ color: Colors.white }}
                            icon={'check_circle_outline'}
                            IconStyle={[{ color: Colors.white }, style.btnStyle] as any}
                        />
                    </FooterActionArea>
                </ContentArea>
            </ModalContainer>
        </Modal>
    )
}

export default ColorPickerModal

const style = StyleSheet.create({
    btnStyle: {
        fontSize: 24,
        marginRight: 24,
    },
})
