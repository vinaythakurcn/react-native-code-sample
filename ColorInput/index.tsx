import React, { useState } from 'react'
import Button from '../Button'
import ColorPickerModal from './ColorPickerModal'

export interface IColorInput {

    /**
     * Represent the color value as hexcode
     */
    value?: string

    /**
     * Custom Label or Title of the button
     */
    label?: string

    /**
     * Custom Label or Title of the button
     */
    icon?: string

    /**
     * Event to be triggered when a color selected
     */
    onChange: (value: string) => void

    /**
     * Set this property to `true` to disable the button
     */
    disabled?: boolean

    /**
     * Set this property to `true` to set the button readonly
     */
    readOnly?: boolean
}

const ColorInput: React.FC<IColorInput> = (props) => {
    const { value: propValue, label, icon, onChange, ...restProps } = props

    const [value, setValue] = useState<string>(propValue || '')
    const [modalVisible, setModalVisible] = useState(false)

    React.useEffect(() => {
        if (!!propValue) {
            setValue(propValue)
        }
    }, [propValue])

    const onSelect = React.useCallback((color: string) => {
        if (restProps.disabled || restProps.readOnly) return;

        setModalVisible(false)

        if (typeof onChange == 'function') {
            setValue(color)
            onChange(color)
        }
    }, [value, onChange])

    return (
        <>
            <Button
                {...restProps}
                disabled={restProps.disabled || restProps.readOnly}
                icon={icon || 'wb-icon-pipette-full'}
                title={label || 'Pick a color'}
                onPress={() => setModalVisible(true)}
                style={{ backgroundColor: value }}
            />
            <ColorPickerModal
                isOpen={modalVisible}
                onClose={() => setModalVisible(false)}
                onChange={(value: string) => onSelect(value)}
            />
        </>
    )
}

const areEqual = (prevProps: IColorInput, nextProps: IColorInput) => {
    return (
      prevProps.value === nextProps.value &&
      prevProps.label === nextProps.label &&
      prevProps.icon === nextProps.icon &&
      prevProps.disabled === nextProps.disabled &&
      prevProps.readOnly === nextProps.readOnly
    );
  };

export default React.memo(ColorInput, areEqual)
