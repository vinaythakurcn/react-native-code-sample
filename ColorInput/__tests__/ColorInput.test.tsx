import * as React from 'react'
import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer'
import ColorInput, {IColorInput} from '../'

const defaultProps: IColorInput = {
    value: '#FF0000',
    onChange: () => {},
}

const setup = (props = {}) => <ColorInput {...defaultProps} {...props} />

describe('<ColorInput />', () => {
    test('Exported correctly', () => {
        expect(setup()).toBeDefined()
    })

    test('Render without failed ', () => {
        const tree = renderer.create(setup()).toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('Verifying props ', () => {
        const tree = renderer.create(setup())
        const {
            root: { props: renderedProp },
        } = tree
        expect(renderedProp.value).toStrictEqual(defaultProps.value)

    })

    test('should exists label when passed ', () => {
        const { queryByText } = render(setup({label: 'some label'}));
        const labelField = queryByText('some label')
        expect(labelField).toBeTruthy()

    })
})
