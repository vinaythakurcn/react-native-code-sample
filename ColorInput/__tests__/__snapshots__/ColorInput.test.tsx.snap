// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorInput /> Render without failed  1`] = `
Array [
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF0000",
        "borderColor": "#DBDBDB",
        "borderRadius": 24,
        "borderWidth": 1.5,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginHorizontal": 15,
          "marginVertical": 4,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "marginHorizontal": 3,
            },
            undefined,
          ]
        }
      >
        <Text>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#3A3F55",
                  "fontSize": 18,
                },
                Object {},
                Object {
                  "fontFamily": "wb-icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            î¦ª
          </Text>
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Poppins-Regular",
              "fontSize": 16,
            },
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Poppins-SemiBold",
                "paddingHorizontal": 8,
                "paddingVertical": 8,
              },
              false,
              Object {
                "color": "#656464",
              },
              undefined,
              undefined,
            ],
          ]
        }
        testID="button-text"
      >
        Pick a color
      </Text>
    </View>
  </View>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={false}
    visible={false}
  />,
]
`;
